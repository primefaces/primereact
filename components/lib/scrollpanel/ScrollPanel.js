import * as React from 'react';
import { useMountEffect, useUnmountEffect } from '../hooks/Hooks';
import { classNames, DomHandler } from '../utils/Utils';
import { ScrollPanelBase } from './ScrollPanelBase';

export const ScrollPanel = React.forwardRef((inProps, ref) => {
    const props = ScrollPanelBase.getProps(inProps);

    const containerRef = React.useRef(null);
    const contentRef = React.useRef(null);
    const xBarRef = React.useRef(null);
    const yBarRef = React.useRef(null);
    const isXBarClicked = React.useRef(false);
    const isYBarClicked = React.useRef(false);
    const lastPageX = React.useRef(null);
    const lastPageY = React.useRef(null);
    const scrollXRatio = React.useRef(null);
    const scrollYRatio = React.useRef(null);
    const frame = React.useRef(null);
    const initialized = React.useRef(false);

    const calculateContainerHeight = () => {
        const containerStyles = getComputedStyle(containerRef.current);
        const xBarStyles = getComputedStyle(xBarRef.current);
        const pureContainerHeight = DomHandler.getHeight(containerRef.current) - parseInt(xBarStyles['height'], 10);

        if (containerStyles['max-height'] !== 'none' && pureContainerHeight === 0) {
            if (contentRef.current.offsetHeight + parseInt(xBarStyles['height'], 10) > parseInt(containerStyles['max-height'], 10)) {
                containerRef.current.style.height = containerStyles['max-height'];
            } else {
                containerRef.current.style.height =
                    contentRef.current.offsetHeight + parseFloat(containerStyles.paddingTop) + parseFloat(containerStyles.paddingBottom) + parseFloat(containerStyles.borderTopWidth) + parseFloat(containerStyles.borderBottomWidth) + 'px';
            }
        }
    };

    const moveBar = () => {
        // horizontal scroll
        const totalWidth = contentRef.current.scrollWidth;
        const ownWidth = contentRef.current.clientWidth;
        const bottom = (containerRef.current.clientHeight - xBarRef.current.clientHeight) * -1;

        scrollXRatio.current = ownWidth / totalWidth;

        // vertical scroll
        const totalHeight = contentRef.current.scrollHeight;
        const ownHeight = contentRef.current.clientHeight;
        const right = (containerRef.current.clientWidth - yBarRef.current.clientWidth) * -1;

        scrollYRatio.current = ownHeight / totalHeight;

        frame.current = window.requestAnimationFrame(() => {
            if (scrollXRatio.current >= 1) {
                DomHandler.addClass(xBarRef.current, 'p-scrollpanel-hidden');
            } else {
                DomHandler.removeClass(xBarRef.current, 'p-scrollpanel-hidden');
                xBarRef.current.style.cssText = 'width:' + Math.max(scrollXRatio.current * 100, 10) + '%; left:' + (contentRef.current.scrollLeft / totalWidth) * 100 + '%;bottom:' + bottom + 'px;';
            }

            if (scrollYRatio.current >= 1) {
                DomHandler.addClass(yBarRef.current, 'p-scrollpanel-hidden');
            } else {
                DomHandler.removeClass(yBarRef.current, 'p-scrollpanel-hidden');
                yBarRef.current.style.cssText = 'height:' + Math.max(scrollYRatio.current * 100, 10) + '%; top: calc(' + (contentRef.current.scrollTop / totalHeight) * 100 + '% - ' + xBarRef.current.clientHeight + 'px);right:' + right + 'px;';
            }
        });
    };

    const onYBarMouseDown = (event) => {
        isYBarClicked.current = true;
        lastPageY.current = event.pageY;
        DomHandler.addClass(yBarRef.current, 'p-scrollpanel-grabbed');
        DomHandler.addClass(document.body, 'p-scrollpanel-grabbed');

        document.addEventListener('mousemove', onDocumentMouseMove);
        document.addEventListener('mouseup', onDocumentMouseUp);
        event.preventDefault();
    };

    const onXBarMouseDown = (event) => {
        isXBarClicked.current = true;
        lastPageX.current = event.pageX;
        DomHandler.addClass(xBarRef.current, 'p-scrollpanel-grabbed');
        DomHandler.addClass(document.body, 'p-scrollpanel-grabbed');

        document.addEventListener('mousemove', onDocumentMouseMove);
        document.addEventListener('mouseup', onDocumentMouseUp);
        event.preventDefault();
    };

    const onDocumentMouseMove = (event) => {
        if (isXBarClicked.current) {
            onMouseMoveForXBar(event);
        } else if (isYBarClicked.current) {
            onMouseMoveForYBar(event);
        } else {
            onMouseMoveForXBar(event);
            onMouseMoveForYBar(event);
        }
    };

    const onMouseMoveForXBar = (event) => {
        const deltaX = event.pageX - lastPageX.current;

        lastPageX.current = event.pageX;

        frame.current = window.requestAnimationFrame(() => {
            contentRef.current.scrollLeft += deltaX / scrollXRatio.current;
        });
    };

    const onMouseMoveForYBar = (event) => {
        const deltaY = event.pageY - lastPageY.current;

        lastPageY.current = event.pageY;

        frame.current = window.requestAnimationFrame(() => {
            contentRef.current.scrollTop += deltaY / scrollYRatio.current;
        });
    };

    const onDocumentMouseUp = (event) => {
        DomHandler.removeClass(yBarRef.current, 'p-scrollpanel-grabbed');
        DomHandler.removeClass(xBarRef.current, 'p-scrollpanel-grabbed');
        DomHandler.removeClass(document.body, 'p-scrollpanel-grabbed');

        document.removeEventListener('mousemove', onDocumentMouseMove);
        document.removeEventListener('mouseup', onDocumentMouseUp);
        isXBarClicked.current = false;
        isYBarClicked.current = false;
    };

    const refresh = () => {
        moveBar();
    };

    useMountEffect(() => {
        moveBar();
        window.addEventListener('resize', moveBar);
        calculateContainerHeight();
        initialized.current = true;
    });

    useUnmountEffect(() => {
        if (initialized.current) {
            window.removeEventListener('resize', moveBar);
        }

        if (frame.current) {
            window.cancelAnimationFrame(frame.current);
        }
    });

    React.useImperativeHandle(ref, () => ({
        props,
        refresh,
        getElement: () => containerRef.current,
        getContent: () => contentRef.current,
        getXBar: () => xBarRef.current,
        getYBar: () => yBarRef.current
    }));

    const otherProps = ScrollPanelBase.getOtherProps(props);
    const className = classNames('p-scrollpanel p-component', props.className);

    return (
        <div ref={containerRef} id={props.id} className={className} style={props.style} {...otherProps}>
            <div className="p-scrollpanel-wrapper">
                <div ref={contentRef} className="p-scrollpanel-content" onScroll={moveBar} onMouseEnter={moveBar}>
                    {props.children}
                </div>
            </div>
            <div ref={xBarRef} className="p-scrollpanel-bar p-scrollpanel-bar-x" onMouseDown={onXBarMouseDown}></div>
            <div ref={yBarRef} className="p-scrollpanel-bar p-scrollpanel-bar-y" onMouseDown={onYBarMouseDown}></div>
        </div>
    );
});

ScrollPanel.displayName = 'ScrollPanel';
