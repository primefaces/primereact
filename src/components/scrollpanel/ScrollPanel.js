import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { DomHandler, classNames } from '../utils/Utils';

export class ScrollPanel extends Component {

    static defaultProps = {
        id: null,
        style: null,
        className: null
    }

    static propTypes = {
        id: PropTypes.string,
        style: PropTypes.object,
        className: PropTypes.string
    };

    constructor(props) {
        super(props);

        this.moveBar = this.moveBar.bind(this);
        this.onXBarMouseDown = this.onXBarMouseDown.bind(this);
        this.onYBarMouseDown = this.onYBarMouseDown.bind(this);

        this.onDocumentMouseMove = this.onDocumentMouseMove.bind(this);
        this.onDocumentMouseUp = this.onDocumentMouseUp.bind(this);
    }

    calculateContainerHeight() {
        let containerStyles = getComputedStyle(this.container),
        xBarStyles = getComputedStyle(this.xBar),
        pureContainerHeight = DomHandler.getHeight(this.container) - parseInt(xBarStyles['height'], 10);

        if (containerStyles['max-height'] !== "none" && pureContainerHeight === 0) {
            if(this.content.offsetHeight + parseInt(xBarStyles['height'], 10) > parseInt(containerStyles['max-height'], 10)) {
                this.container.style.height = containerStyles['max-height'];
            }
            else {
                this.container.style.height = this.content.offsetHeight + parseFloat(containerStyles.paddingTop) + parseFloat(containerStyles.paddingBottom) + parseFloat(containerStyles.borderTopWidth) + parseFloat(containerStyles.borderBottomWidth) + "px";
            }
        }
    }

    moveBar() {
        /* horizontal scroll */
        let totalWidth = this.content.scrollWidth;
        let ownWidth = this.content.clientWidth;
        let bottom = (this.container.clientHeight - this.xBar.clientHeight) * -1;

        this.scrollXRatio = ownWidth / totalWidth;

        /* vertical scroll */
        let totalHeight = this.content.scrollHeight;
        let ownHeight = this.content.clientHeight;
        let right = (this.container.clientWidth - this.yBar.clientWidth) * -1;

        this.scrollYRatio = ownHeight / totalHeight;

        this.frame = this.requestAnimationFrame(() => {
            if (this.scrollXRatio >= 1) {
                DomHandler.addClass(this.xBar, 'p-scrollpanel-hidden');
            }
            else {
                DomHandler.removeClass(this.xBar, 'p-scrollpanel-hidden');
                this.xBar.style.cssText = 'width:' + Math.max(this.scrollXRatio * 100, 10) + '%; left:' + (this.content.scrollLeft / totalWidth) * 100 + '%;bottom:' + bottom + 'px;';
            }

            if (this.scrollYRatio >= 1) {
                DomHandler.addClass(this.yBar, 'p-scrollpanel-hidden');
            }
            else {
                DomHandler.removeClass(this.yBar, 'p-scrollpanel-hidden');
                this.yBar.style.cssText = 'height:' + Math.max(this.scrollYRatio * 100, 10) + '%; top: calc(' + (this.content.scrollTop / totalHeight) * 100 + '% - ' + this.xBar.clientHeight + 'px);right:' + right + 'px;';
            }
        });
    }

    onYBarMouseDown(e) {
        this.isYBarClicked = true;
        this.lastPageY = e.pageY;
        DomHandler.addClass(this.yBar, 'p-scrollpanel-grabbed');
        DomHandler.addClass(document.body, 'p-scrollpanel-grabbed');

        document.addEventListener('mousemove', this.onDocumentMouseMove);
        document.addEventListener('mouseup', this.onDocumentMouseUp);
        e.preventDefault();
    }

    onXBarMouseDown(e) {
        this.isXBarClicked = true;
        this.lastPageX = e.pageX;
        DomHandler.addClass(this.xBar, 'p-scrollpanel-grabbed');
        DomHandler.addClass(document.body, 'p-scrollpanel-grabbed');

        document.addEventListener('mousemove', this.onDocumentMouseMove);
        document.addEventListener('mouseup', this.onDocumentMouseUp);
        e.preventDefault();
    }

    onDocumentMouseMove(e) {
        if(this.isXBarClicked) {
            this.onMouseMoveForXBar(e);
        }
        else if(this.isYBarClicked) {
            this.onMouseMoveForYBar(e);
        }
        else {
            this.onMouseMoveForXBar(e);
            this.onMouseMoveForYBar(e);
        }

    }

    onMouseMoveForXBar(e) {
        let deltaX = e.pageX - this.lastPageX;
        this.lastPageX = e.pageX;

        this.frame = this.requestAnimationFrame(() => {
            this.content.scrollLeft += deltaX / this.scrollXRatio;
        });
    }

    onMouseMoveForYBar(e) {
        let deltaY = e.pageY - this.lastPageY;
        this.lastPageY = e.pageY;

        this.frame = this.requestAnimationFrame(() => {
            this.content.scrollTop += deltaY / this.scrollYRatio;
        });
    }

    onDocumentMouseUp(e) {
        DomHandler.removeClass(this.yBar, 'p-scrollpanel-grabbed');
        DomHandler.removeClass(this.xBar, 'p-scrollpanel-grabbed');
        DomHandler.removeClass(document.body, 'p-scrollpanel-grabbed');

        document.removeEventListener('mousemove', this.onDocumentMouseMove);
        document.removeEventListener('mouseup', this.onDocumentMouseUp);
        this.isXBarClicked = false;
        this.isYBarClicked = false;
    }

    requestAnimationFrame(f) {
        let frame = window.requestAnimationFrame || this.timeoutFrame;
        return frame(f);
    }

    refresh() {
        this.moveBar();
    }

    componentDidMount() {
        this.moveBar();
        this.moveBar = this.moveBar.bind(this);

        window.addEventListener('resize', this.moveBar);

        this.calculateContainerHeight();
        this.initialized = true;
    }

    componentWillUnmount() {
        if (this.initialized) {
            window.removeEventListener('resize', this.moveBar);
        }

        if (this.frame) {
            window.cancelAnimationFrame(this.frame);
        }
    }

    render() {
        let className = classNames('p-scrollpanel p-component', this.props.className);

        return (
            <div ref={(el) => this.container = el} id={this.props.id} className={className} style={this.props.style}>
                <div className="p-scrollpanel-wrapper">
                    <div ref={(el) => this.content = el} className="p-scrollpanel-content" onScroll={this.moveBar} onMouseEnter={this.moveBar}>
                        {this.props.children}
                    </div>
                </div>
                <div ref={(el) => this.xBar = el} className="p-scrollpanel-bar p-scrollpanel-bar-x" onMouseDown={this.onXBarMouseDown}></div>
                <div ref={(el) => this.yBar = el} className="p-scrollpanel-bar p-scrollpanel-bar-y" onMouseDown={this.onYBarMouseDown}></div>
            </div>
        );
    }
}
