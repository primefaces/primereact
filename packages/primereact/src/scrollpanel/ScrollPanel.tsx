'use client';
import { Component, withComponent } from '@primereact/core/component';
import { useScrollPanel } from '@primereact/headless/scrollpanel';
import { styles } from '@primereact/styles/scrollpanel';
import { mergeProps } from '@primeuix/utils';
import * as React from 'react';
import { defaultProps } from './ScrollPanel.props';
import { ScrollPanelContent } from './content';

export const ScrollPanel = withComponent({
    defaultProps,
    styles,
    setup: (instance) => {
        const scrollPanel = useScrollPanel(instance.inProps);

        return scrollPanel;
    },
    render: (instance) => {
        const {
            id,
            props,
            ptmi,
            ptm,
            cx,
            // element refs
            elementRef
        } = instance;

        const { contentId, xBarRef, yBarRef, lastScrollLeft, lastScrollTop, onXBarMouseDown, onYBarMouseDown, onFocus, onBlur, onKeyDown, onKeyUp } = instance;

        const rootProps = mergeProps(
            {
                id,
                className: cx('root')
            },
            ptmi('root')
        );

        const createContentContainer = () => {
            const contentContainerProps = mergeProps(
                {
                    className: cx('contentContainer')
                },
                ptm('contentContainer')
            );

            return <div {...contentContainerProps}>{props.children}</div>;
        };

        const createXBar = () => {
            const xBarProps = mergeProps(
                {
                    className: cx('barx'),
                    tabIndex: 0,
                    role: 'scrollbar',
                    'aria-orientation': 'horizontal',
                    'aria-controls': contentId,
                    'aria-valuenow': lastScrollLeft,
                    onMouseDown: onXBarMouseDown,
                    onKeyDown,
                    onKeyUp,
                    onFocus,
                    onBlur,
                    'data-pc-group-section': 'bar'
                },
                ptm('xBar')
            );

            return <div ref={xBarRef} {...xBarProps} />;
        };

        const createYBar = () => {
            const yBarProps = mergeProps(
                {
                    className: cx('bary'),
                    tabIndex: 0,
                    role: 'scrollbar',
                    'aria-orientation': 'vertical',
                    'aria-controls': contentId,
                    'aria-valuenow': lastScrollTop,
                    onMouseDown: onYBarMouseDown,
                    onKeyDown,
                    onKeyUp,
                    onFocus,
                    onBlur,
                    'data-pc-group-section': 'bar'
                },
                ptm('xBar')
            );

            return <div ref={yBarRef} {...yBarProps} />;
        };

        const contentContainer = createContentContainer();
        const xBar = createXBar();
        const yBar = createYBar();

        return (
            <Component as={props.as || 'div'} {...rootProps} ref={elementRef}>
                {contentContainer}
                {xBar}
                {yBar}
            </Component>
        );
    },
    components: {
        Content: ScrollPanelContent
    }
});
