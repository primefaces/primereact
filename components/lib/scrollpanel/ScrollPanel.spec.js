import '@testing-library/jest-dom';
import { act, render } from '@testing-library/react';
import * as React from 'react';
import { snapshot } from '../../test';
import { ScrollPanel } from './ScrollPanel';

describe('ScrollPanel', () => {
    snapshot(
        <>
            <ScrollPanel id="scrollpanel" style={{ width: '100%', height: '200px' }} className="jest">
                <div style={{ padding: '1em', lineHeight: '1.5' }}>
                    The story begins as Don Vito Corleone, the head of a New York Mafia family, oversees his daughter's wedding. His beloved son Michael has just come home from the war, but does not intend to become part of his father's business.
                    Through Michael's life the nature of the family business becomes clear. The business of the family is just like the head of the family, kind and benevolent to those who give respect, but given to ruthless violence whenever
                    anything stands against the good of the family. The story begins as Don Vito Corleone, the head of a New York Mafia family, oversees his daughter's wedding. His beloved son Michael has just come home from the war, but does not
                    intend to become part of his father's business. Through Michael's life the nature of the family business becomes clear. The business of the family is just like the head of the family, kind and benevolent to those who give respect,
                    but given to ruthless violence whenever anything stands against the good of the family.
                </div>
            </ScrollPanel>
        </>,
        'ScrollPanel full settings'
    );
    test('refresh method', async () => {
        // Arrange
        const ref = React.createRef();
        const { container } = render(
            <ScrollPanel id="scrollpanel" style={{ width: '100%', height: '200px' }} className="jest" ref={ref}>
                <div style={{ padding: '1em', lineHeight: '1.5' }}>
                    The story begins as Don Vito Corleone, the head of a New York Mafia family, oversees his daughter's wedding. His beloved son Michael has just come home from the war, but does not intend to become part of his father's business.
                    Through Michael's life the nature of the family business becomes clear. The business of the family is just like the head of the family, kind and benevolent to those who give respect, but given to ruthless violence whenever
                    anything stands against the good of the family. The story begins as Don Vito Corleone, the head of a New York Mafia family, oversees his daughter's wedding. His beloved son Michael has just come home from the war, but does not
                    intend to become part of his father's business. Through Michael's life the nature of the family business becomes clear. The business of the family is just like the head of the family, kind and benevolent to those who give respect,
                    but given to ruthless violence whenever anything stands against the good of the family.
                </div>
            </ScrollPanel>
        );

        // Act
        act(() => {
            ref.current.refresh();
            ref.current.getElement();
            ref.current.getContent();
            ref.current.getXBar();
            ref.current.getYBar();
        });

        // Assert
        expect(container).toMatchSnapshot();
    });
});
