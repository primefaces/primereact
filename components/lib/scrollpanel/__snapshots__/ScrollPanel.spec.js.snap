// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollPanel ScrollPanel full settings 1`] = `
<div>
  <div
    class="p-scrollpanel p-component jest"
    id="scrollpanel"
    style="width: 100%; height: 200px;"
  >
    <div
      class="p-scrollpanel-wrapper"
    >
      <div
        class="p-scrollpanel-content"
      >
        <div
          style="padding: 1em; line-height: 1.5;"
        >
          The story begins as Don Vito Corleone, the head of a New York Mafia family, oversees his daughter's wedding. His beloved son Michael has just come home from the war, but does not intend to become part of his father's business. Through Michael's life the nature of the family business becomes clear. The business of the family is just like the head of the family, kind and benevolent to those who give respect, but given to ruthless violence whenever anything stands against the good of the family. The story begins as Don Vito Corleone, the head of a New York Mafia family, oversees his daughter's wedding. His beloved son Michael has just come home from the war, but does not intend to become part of his father's business. Through Michael's life the nature of the family business becomes clear. The business of the family is just like the head of the family, kind and benevolent to those who give respect, but given to ruthless violence whenever anything stands against the good of the family.
        </div>
      </div>
    </div>
    <div
      class="p-scrollpanel-bar p-scrollpanel-bar-x"
    />
    <div
      class="p-scrollpanel-bar p-scrollpanel-bar-y"
    />
  </div>
</div>
`;

exports[`ScrollPanel refresh method 1`] = `
<div>
  <div
    class="p-scrollpanel p-component jest"
    id="scrollpanel"
    style="width: 100%; height: 200px;"
  >
    <div
      class="p-scrollpanel-wrapper"
    >
      <div
        class="p-scrollpanel-content"
      >
        <div
          style="padding: 1em; line-height: 1.5;"
        >
          The story begins as Don Vito Corleone, the head of a New York Mafia family, oversees his daughter's wedding. His beloved son Michael has just come home from the war, but does not intend to become part of his father's business. Through Michael's life the nature of the family business becomes clear. The business of the family is just like the head of the family, kind and benevolent to those who give respect, but given to ruthless violence whenever anything stands against the good of the family. The story begins as Don Vito Corleone, the head of a New York Mafia family, oversees his daughter's wedding. His beloved son Michael has just come home from the war, but does not intend to become part of his father's business. Through Michael's life the nature of the family business becomes clear. The business of the family is just like the head of the family, kind and benevolent to those who give respect, but given to ruthless violence whenever anything stands against the good of the family.
        </div>
      </div>
    </div>
    <div
      class="p-scrollpanel-bar p-scrollpanel-bar-x"
    />
    <div
      class="p-scrollpanel-bar p-scrollpanel-bar-y"
    />
  </div>
</div>
`;
